examples/example_customdns_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) MongoDB, Inc. 2022-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package examples
+
+import (
+	"context"
+	"log"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func resolve(ctx context.Context, cache *dnsCache, in *dns.Conn, out *dns.Conn) {
+	for ctx.Err() == nil {
+		q, err := in.ReadMsg()
+		if err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in ReadMsg: %v", err)
+		}
+		if len(q.Question) != 1 {
+			// Multiple questions in a single query is not actually used in real life.
+			continue
+		}
+
+		a, err := func() (*dns.Msg, error) {
+			cache.lock.Lock()
+			defer cache.lock.Unlock()
+
+			now := time.Now()
+			if rr, ok := cache.records[q.Question[0]]; ok && rr.exp.After(now) {
+				a := new(dns.Msg)
+				a.SetReply(q)
+				a.Compress = false
+				a.Answer = append(a.Answer, rr.record)
+				return a, nil
+			}
+
+			err := out.WriteMsg(q)
+			if err != nil {
+				return nil, err
+			}
+
+			m, err := out.ReadMsg()
+			if err != nil {
+				return nil, err
+			}
+
+			l := len(m.Answer)
+			for i, q := range m.Question {
+				if i >= l {
+					break
+				}
+				a := m.Answer[i]
+				cache.records[q] = &RR{
+					a,
+					now.Add(time.Second * time.Duration(a.Header().Ttl)),
+				}
+			}
+			return m, nil
+		}()
+		if err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in record retrieval: %v", err)
+		}
+
+		if err := in.WriteMsg(a); err != nil {
+			// TODO: Handle error.
+			log.Fatalf("Unhandled error in WriteMsg: %v", err)
+		}
+	}
+}
+
+type RR struct {
+	record dns.RR
+	exp    time.Time
+}
+
+type dnsCache struct {
+	records map[dns.Question]*RR
+	lock    *sync.Mutex
+}
+
+type dialer struct {
+	*net.Dialer
+	cache *dnsCache
+}
+
+func NewDialer() dialer {
+	cache := &dnsCache{
+		records: make(map[dns.Question]*RR),
+		lock:    &sync.Mutex{},
+	}
+	return dialer{
+		Dialer: &net.Dialer{
+			Resolver: &net.Resolver{
+				PreferGo: true,
+				Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+					outConn, err := net.Dial(network, address)
+					conn, inConn := net.Pipe()
+					if err == nil {
+						go resolve(ctx, cache, &dns.Conn{Conn: inConn}, &dns.Conn{Conn: outConn})
+					}
+					return conn, err
+				},
+			},
+		},
+		cache: cache,
+	}
+}
+
+func TestCustomDialer(t *testing.T) {
+	client, err := mongo.NewClient(options.Client().ApplyURI("mongodb://testurl:27017").SetDialer(NewDialer()))
```

Comment:
When I run this locally it dials until it reaches a timeout via a heartbeat: 

> connection() error occurred during connection handshake: dial tcp: lookup testurl on 127.0.0.1:53: no such host

 If I switch to `"mongodb://localhost:27017"`/remap `testurl` to localhost (via /etc/hosts), then the connection won't use the custom dialer. I think this is because [under the hood](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/connection.go#L189) we use `DialContext`?

Reply:
I observed the same behavior. However, I was not aware of the `DialContext` in connection.go. I thought it was caused by using the special host, `localhost`. I can dig deeper.

Reply:
@qingyang-hu I believe you're correct that `localhost` is a special case because it's defined in `/etc/hosts`, which bypasses the whole "dial a connection to a DNS resolver" flow (i.e. never calls `Resolver.Dial`). Consider using one of the DNS names used in the [Initial DNS Seedlist Discovery spec tests](https://github.com/mongodb/specifications/tree/master/source/initial-dns-seedlist-discovery/tests), like `localhost.test.build.10gen.cc`, which is an a registered DNS name that resolves to `127.0.0.1`.

Neither the [DefaultResolver](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/net/lookup.go;l=117) nor the overridden `Resolver` here use the `mongo.Client` dialer to dial the actual DNS server, so the `DialContext` call in `connection.go` should not be relevant.